File: wrappers/go/ccl/providers.go
package ccl

// Optional chain-data provider helpers.
//
// QuickTx.Build is offline by design: the caller supplies UTXOs and protocol parameters (and, for
// Plutus, execution units). These helpers are an optional convenience that fetch those inputs from a
// chain-data backend over HTTP, returning them in the shape Build already accepts — so the native
// library stays offline and provider-free, and the helpers are pure wrapper-side code using the
// standard net/http client.
//
// A provider implements two methods:
//
//	Utxos(address)     -> all UTXOs at the address (no selection — the bridge selects)
//	ProtocolParams()   -> protocol parameters
//
// Use one directly, or via QuickTxApi.BuildWithProvider:
//
//	provider := ccl.NewBlockfrostProvider(projectID, "preprod") // or ccl.NewYaciProvider("")
//	result, err := bridge.QuickTx.BuildWithProvider(yaml, provider, senderAddress)

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
)

// ChainDataProvider fetches the chain data QuickTx.Build needs. Implement it to plug in any backend
// (Blockfrost, Koios, Ogmios, Yaci DevKit, ...).
type ChainDataProvider interface {
	// Utxos returns all UTXOs at the address (the bridge selects internally; no selection needed).
	Utxos(address string) ([]map[string]interface{}, error)
	// ProtocolParams returns the current protocol parameters.
	ProtocolParams() (map[string]interface{}, error)
}

func httpGetJSON(url string, headers map[string]string, out interface{}) error {
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return err
	}
	for k, v := range headers {
		req.Header.Set(k, v)
	}
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("GET %s failed: HTTP %d: %s", url, resp.StatusCode, string(body))
	}
	return json.NewDecoder(resp.Body).Decode(out)
}

// YaciProvider is a ChainDataProvider backed by Yaci DevKit / yaci-store (Blockfrost-style REST).
// Its UTXO and protocol-parameter responses are already in the shape Build expects.
type YaciProvider struct {
	BaseURL string
}

const defaultYaciURL = "http://localhost:10000/local-cluster/api"

// NewYaciProvider returns a YaciProvider for the given base URL; pass "" for the local DevKit cluster.
func NewYaciProvider(baseURL string) *YaciProvider {
	if baseURL == "" {
		baseURL = defaultYaciURL
	}
	return &YaciProvider{BaseURL: strings.TrimRight(baseURL, "/")}
}

func (p *YaciProvider) Utxos(address string) ([]map[string]interface{}, error) {
	var utxos []map[string]interface{}
	err := httpGetJSON(fmt.Sprintf("%s/addresses/%s/utxos", p.BaseURL, address), nil, &utxos)
	return utxos, err
}

func (p *YaciProvider) ProtocolParams() (map[string]interface{}, error) {
	var pp map[string]interface{}
	err := httpGetJSON(p.BaseURL+"/epochs/parameters", nil, &pp)
	return pp, err
}

var blockfrostNetworkURLs = map[string]string{
	"mainnet": "https://cardano-mainnet.blockfrost.io/api/v0",
	"preprod": "https://cardano-preprod.blockfrost.io/api/v0",
	"preview": "https://cardano-preview.blockfrost.io/api/v0",
}

// BlockfrostProvider is a ChainDataProvider backed by the Blockfrost API. UTXOs are paginated 100 per
// page, and Blockfrost omits the owning address on each UTXO so it is injected.
type BlockfrostProvider struct {
	BaseURL   string
	ProjectID string
}

// NewBlockfrostProvider returns a provider for the given network ("mainnet"/"preprod"/"preview").
// Use NewBlockfrostProviderURL to point at a custom base URL.
func NewBlockfrostProvider(projectID, network string) (*BlockfrostProvider, error) {
	baseURL, ok := blockfrostNetworkURLs[network]
	if !ok {
		return nil, fmt.Errorf("unknown network %q; use NewBlockfrostProviderURL", network)
	}
	return NewBlockfrostProviderURL(projectID, baseURL), nil
}

// NewBlockfrostProviderURL returns a provider pointed at an explicit base URL (e.g. self-hosted).
func NewBlockfrostProviderURL(projectID, baseURL string) *BlockfrostProvider {
	return &BlockfrostProvider{BaseURL: strings.TrimRight(baseURL, "/"), ProjectID: projectID}
}

func (p *BlockfrostProvider) headers() map[string]string {
	return map[string]string{"project_id": p.ProjectID}
}

func (p *BlockfrostProvider) Utxos(address string) ([]map[string]interface{}, error) {
	var out []map[string]interface{}
	for page := 1; ; page++ {
		var items []map[string]interface{}
		url := fmt.Sprintf("%s/addresses/%s/utxos?count=100&page=%d", p.BaseURL, address, page)
		if err := httpGetJSON(url, p.headers(), &items); err != nil {
			return nil, err
		}
		if len(items) == 0 {
			break
		}
		for _, u := range items {
			// Blockfrost omits the owning address on each UTXO; Build needs it.
			if _, ok := u["address"]; !ok {
				u["address"] = address
			}
			out = append(out, u)
		}
		if len(items) < 100 {
			break
		}
	}
	return out, nil
}

func (p *BlockfrostProvider) ProtocolParams() (map[string]interface{}, error) {
	// Blockfrost's parameters are a superset of CCL's ProtocolParams; the native lib ignores
	// unknown fields, so the response passes through unchanged.
	var pp map[string]interface{}
	err := httpGetJSON(p.BaseURL+"/epochs/latest/parameters", p.headers(), &pp)
	return pp, err
}

// BuildWithProvider fetches chain data from the provider and builds, in one call — composing
// provider.Utxos(sender) + provider.ProtocolParams() with Build. The bridge stays offline; this only
// moves the optional HTTP fetch into wrapper code.
func (q *QuickTxApi) BuildWithProvider(yaml string, provider ChainDataProvider, sender string, execUnits ...interface{}) (*TxResult, error) {
	utxos, err := provider.Utxos(sender)
	if err != nil {
		return nil, fmt.Errorf("provider utxos: %w", err)
	}
	pp, err := provider.ProtocolParams()
	if err != nil {
		return nil, fmt.Errorf("provider protocol params: %w", err)
	}
	return q.Build(yaml, utxos, pp, execUnits...)
}

File: wrappers/python/ccl/providers.py
"""Optional chain-data provider helpers.

``QuickTx.build`` is offline by design: the caller supplies UTXOs and protocol parameters (and, for
Plutus, execution units). These helpers are an *optional* convenience that fetch those inputs from a
chain-data backend over HTTP, returning them in the exact shape ``build`` already accepts — so the
native library stays offline and provider-free, and the helpers are pure wrapper-side code using
only stdlib ``urllib``.

A provider implements two methods:

    utxos(address)        -> list of UTXO dicts at the address (no selection — the bridge selects)
    protocol_params()     -> protocol parameters dict

Use one directly, or via the ``QuickTx.build_with_provider`` convenience::

    from ccl import CclLib
    from ccl.providers import BlockfrostProvider

    lib = CclLib()
    provider = BlockfrostProvider(project_id, network="preprod")   # or YaciProvider() for DevKit
    result = lib.quicktx.build_with_provider(txplan_yaml, provider, sender_address)
"""
import json
import urllib.request
import urllib.error


class ChainDataProvider:
    """Interface for fetching the chain data ``QuickTx.build`` needs.

    Implement ``utxos`` and ``protocol_params`` to plug in any backend (Blockfrost, Koios, Ogmios,
    Yaci DevKit, ...). Both must return data in the shapes ``build`` accepts.
    """

    def utxos(self, address):
        """Return all UTXOs at ``address`` as a list of dicts (CCL ``Utxo`` shape)."""
        raise NotImplementedError

    def protocol_params(self):
        """Return the current protocol parameters as a dict (CCL ``ProtocolParams`` shape)."""
        raise NotImplementedError


def _http_get_json(url, headers=None, timeout=30):
    req = urllib.request.Request(url, method="GET", headers=headers or {})
    try:
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            return json.loads(resp.read())
    except urllib.error.HTTPError as e:
        body = e.read().decode("utf-8", "replace")
        raise RuntimeError(f"GET {url} failed: HTTP {e.code}: {body}") from None


class YaciProvider(ChainDataProvider):
    """Chain-data provider backed by Yaci DevKit / yaci-store (Blockfrost-style REST).

    Defaults to the local DevKit cluster the integration tests use. The UTXO and protocol-parameter
    responses are already in the shape ``build`` expects, so they pass through unchanged.
    """

    DEFAULT_URL = "http://localhost:10000/local-cluster/api"

    def __init__(self, base_url=DEFAULT_URL):
        self.base_url = base_url.rstrip("/")

    def utxos(self, address):
        return _http_get_json(f"{self.base_url}/addresses/{address}/utxos")

    def protocol_params(self):
        return _http_get_json(f"{self.base_url}/epochs/parameters")


class BlockfrostProvider(ChainDataProvider):
    """Chain-data provider backed by the Blockfrost API.

    ``network`` selects the default base URL (``mainnet`` / ``preprod`` / ``preview``); pass
    ``base_url`` to override (e.g. a self-hosted Blockfrost). Requires a project id. UTXOs are
    paginated 100 per page; Blockfrost omits the owning address on each UTXO, so it is injected.
    """

    _NETWORK_URLS = {
        "mainnet": "https://cardano-mainnet.blockfrost.io/api/v0",
        "preprod": "https://cardano-preprod.blockfrost.io/api/v0",
        "preview": "https://cardano-preview.blockfrost.io/api/v0",
    }

    def __init__(self, project_id, network="mainnet", base_url=None):
        if base_url is None:
            if network not in self._NETWORK_URLS:
                raise ValueError(f"unknown network {network!r}; pass base_url explicitly")
            base_url = self._NETWORK_URLS[network]
        self.base_url = base_url.rstrip("/")
        self._headers = {"project_id": project_id}

    def utxos(self, address):
        out = []
        page = 1
        while True:
            items = _http_get_json(
                f"{self.base_url}/addresses/{address}/utxos?count=100&page={page}",
                headers=self._headers,
            )
            if not items:
                break
            for u in items:
                # Blockfrost omits the owning address on each UTXO; build() needs it.
                u.setdefault("address", address)
                out.append(u)
            if len(items) < 100:
                break
            page += 1
        return out

    def protocol_params(self):
        # Blockfrost's parameters are a superset of CCL's ProtocolParams; the native lib ignores
        # unknown fields, so the response passes through unchanged.
        return _http_get_json(f"{self.base_url}/epochs/latest/parameters", headers=self._headers)
